### ISSUE TITLE
Manually send custom metrics (Agent API or integration with a Metrics Library)

### ISSUE THREAD

**Issue by gregkalapos** [elastic/apm-agent-dotnet#322]
We should offer users the ability to manually send metrics.

I think there would be 2 approaches:
- Integrate with an existing metrics library (e.g. [AppMetrics](https://github.com/AppMetrics/AppMetrics), [Metrics.NET](https://github.com/etishor/Metrics.NET)) 
- Extend the public agent API

1. step:  decide which approach to take 
2. step: implement.

**Comment by oleh-zheleznyak** [comment]
For my project having custom metrics is really important, and it is also available in other APM products on the market. 
It would be really awesome to have this!

From a consumers point of view, in a ASP .NET Core application, I would expect to be able to register a source of `IMetricsProvider` (it could be a list, a factory, etc) with my DI container by using a simple extension method on `IServiceCollection` for example `services.AddElasticApmMetrics(IReadOnlyCollection<IMetricsProvider> metricProviders)`
(That collection/factory could later be injected into a new `MetricsCollector` class ctor)

Without custom metrics key stakeholders in our company will most likely choose another product.

Thanks a lot for the great effort you're doing with APM agent for dotnet!

**Comment by maverix** [comment]
Could this be done using Agent.PayloadSender.QueueMetrics(IMetricSet metrics);

**Comment by maverix** [comment]
Otherwise the IMetricsCollector.MetricsProviders needs to be exposed via a public API

**Comment by maverix** [comment]
https://github.com/workbetter-com-au/apm-agent-dotnet-metrics
Attempted a quick manual 'send metric'.  Couldn't get it to work.  No errors were logged.
I will pull into the project the actual Elastic.Apm to troubleshoot why it didn't work

**Comment by gregkalapos** [comment]
> Couldn't get it to work. No errors were logged.

Let me try to help.

The agent currently sends data to the APM Server when one of these 2 things happen:
- A transaction ends (`Transaction.End()` is called either manually or by auto instrumentation)
- There are at [least 20 events](https://github.com/elastic/apm-agent-dotnet/blob/master/src/Elastic.Apm/Report/PayloadSenderV2.cs#L29) recorded (1 event is e.g. a span, or a metrics with 1 timespan)

In your sample you only record metrics, so probably that's the issue - you need at least 20 of those to make the agent send it to the server. You can either force it by also capturing the transaction (just turn on the agent normally with `UseElasticApm`) - or simply change the code temporarily - or send at least 20 metrics to fill the queue. 

Probably we should add a config setting for this... that'd be a nicer way of doing this.

Also I'd like to point out that in [`PayloadSenderV2`](https://github.com/elastic/apm-agent-dotnet/blob/master/src/Elastic.Apm/Report/PayloadSenderV2.cs) we already do the big part of the work (serialization and sending to the server) on a separate thread, the `payloadSender.QueueMetrics(metrics)` already returns very fast. I'm not sure your `BatchWorker` is needed.

Hope this helps.

**Comment by maverix** [comment]
I think I have identified the issue
https://github.com/elastic/apm-agent-dotnet/blob/master/src/Elastic.Apm/Report/PayloadSenderV2.cs#L151

The payloadsender specifically checks if the metric is of type Metrics.MetricSet
where as I don't have access to the metricset class (Internal)

If this was changed to check for the Interface instead - IMetricSet - Maybe it would work then?

**Comment by maverix** [comment]
I wasn't too familiar with the internal workings of the payloadsender.  I was reproducing the MetricCollector behaviour with the BatchWorker pattern.

**Comment by gregkalapos** [comment]
> If this was changed to check for the Interface instead - IMetricSet - Maybe it would work then?

We should rather make `MetricSet` public - at least that was my plan, its just a plain C# type to transfer data, I think it's a good candidate to use it on the API surface.

**Comment by maverix** [comment]
That would be ideal.  I also hit issues with the TraceParent being internal.  And the internal constructor on the DistributedTracingData object

**Comment by gregkalapos** [comment]
> I also hit issues with the TraceParent being internal. And the internal constructor on the DistributedTracingData object

What usecase was that?

**Comment by maverix** [comment]
I have put together a project in .NET Framework showing how I have integrated with the Elastic.Apm dotnet client
https://github.com/workbetter-com-au/apm-agent-dotnet-framework

**Comment by maverix** [comment]
> > I also hit issues with the TraceParent being internal. And the internal constructor on the DistributedTracingData object
> 
> What usecase was that?

I have put together a project in .NET Framework showing how I have integrated with the Elastic.Apm dotnet client
https://github.com/workbetter-com-au/apm-agent-dotnet-framework

**Comment by SergeyKleyman** [comment]
@andrewharry Thank you - I will definitely take a look at your implementation of  the integrations and consider if and how to make them available with the agent out of the box.

**Comment by marcoregueira** [comment]
I'm closing #708. As suggested this seems a better place for the matter. Just for the record, my suggestion was to allow registration of more (and custom) IMetricsProvider. The OTB functionality is great, but there is no need in hitting you when somebody wants this or that, counter, or even a custom one; I found they're pretty easy to implement if they can be added to the queue...

For me having a pluggable or custom PayloadSenderV2  would be interesting too. I would like to route some of the events to some other system or even add new ones (such as log entries) to the switch in ProcessQueue. (I know that's out of scope for you, as it is paired with the APM service ingestion, but having it open would not harm). Well, by now I have a fork of the agent I can live with, but asking is free :)))

Thank you

**Comment by glucaci** [comment]
I did a PR to have access to `MetricSet` and looking forward to write a `Reporter` for `Metrics.NET`.

In the same time I thought maybe will be a good idea to create a separate package for Metrics, `Elastic.Apm.Metrics` and maybe also for Server communication `Elastic.Apm.Reporter` (`IPayloadSender`). The classes are already splitted on folder level but will be nice to have separate nuget packages for 3rd party integrators libraries.

**Comment by Aaronontheweb** [comment]
Any update on this? We'd be interested in contributing App.Metrics support for Elastic.Apm but it looks like this isn't feasible today.

**Comment by kbalys** [comment]
@gregkalapos - any update on this issue? I am really looking forward to this feature as I would like to send multiple custom metrics using APM.

**Comment by Matiszak** [comment]
Since kibana apm seems ro have support for opentelemetry protocol I think you can try opentelemetry dotnet client white support metrics in beta release and setup otpl peorocol exporter to kibana apm.

**Comment by fuzolan** [comment]
It would be nice to have support for https://docs.microsoft.com/en-us/dotnet/core/diagnostics/event-counters

**Comment by smarandav** [comment]
Is there a demo project that can be used to send a custom metric to the APM Server?

**Comment by kbalys** [comment]
@gregkalapos are there plans in the nearest future to implement this feature?

**Comment by kbalys** [comment]
@gregkalapos, could you share updates or plans regarding this feature?

**Comment by johanknirk** [comment]
I'm also curious if you could get this working by using .net telemetry exporter
https://opentelemetry.io/docs/instrumentation/net/exporters/ 

to setup exporting to OpenTelemetry intake API:
https://www.elastic.co/guide/en/apm/guide/current/api-otlp.html

Anybody tested this ?

**Comment by ArveSystad** [comment]
The OpenTelemetry bridge does not support custom metrics as currently mentioned over on [the docs](https://www.elastic.co/guide/en/apm/agent/dotnet/current/opentelemetry-bridge.html#otel-metrics):

> This bridge only supports the tracing API. The Metrics API is currently not supported.

We have Elastic APM setup and working with all traces and spans, but our custom metrics (from IMeterFactory+Counter<T>) just doesn't seem to be pushed to the stream. The setup code we've used is:

```
builder.Services.AddOpenTelemetry()
                .ConfigureResource(bld => bld.AddService(ServiceName))
                .WithTracing(m => m.AddAspNetCoreInstrumentation().AddOtlpExporter())
                .WithMetrics(m => m.AddAspNetCoreInstrumentation().AddOtlpExporter());
```

Would be really nice if there could either be some form of `AddElasticApmExporter()` or let it be picked up automatically from the metrics APIs somehow. As of now, this is clearly a point where elastic is lacking compared to others. Custom metrics is such a powerful tool.

**Comment by stevejgordon** [comment]
@ArveSystad, Are you using the .NET APM Agent? The setup code you've shown is for the OTel SDK. You should be able to configure the OTLP exporter to send directly to Elastic APM. We are working [on a distro of the SDK](https://github.com/elastic/elastic-otel-dotnet), which eases some of the configuration and compatibility when using the OTel SDK and native OTel instrumentation.

**Comment by ArveSystad** [comment]
@stevejgordon Yes, the APM .NET agent is running nicely and in use indeed - spans and traces pop up in APM as expected.

I'll check out that EDOT-thing (even though it says it's not recommended for production yet). 

Would be nice to be able to push custom metrics through the OTel APIs indeed.

**Comment by stevejgordon** [comment]
@ArveSystad, it's not on our roadmap to pick up custom metrics with the bridge at this time, and we recommend using OpenTelemetry SDKs for such cases. It looks like you already have that in your app. If you configure the OTLP exporter for metrics, you should be able to send those to an Elastic APM Server.

**Comment by stevejgordon** [comment]
Closing this as unplanned but supported for EDOT
